packages/cardpay-reward-programs/cardpay_reward_programs/rules/staking.py

```diff
@@ -0,0 +1,100 @@
+from cardpay_reward_programs.rule import Rule
+import pandas as pd
+
+class Staking(Rule):
+    """
+        This rule rewards users with CARD.cpxd held in a depot in a monthly basis 
+    """
+
+    def __init__(self, core_parameters, user_defined_parameters):
+        super(Staking, self).__init__(core_parameters, user_defined_parameters)
+
+    def set_user_defined_parameters(
+        self,
+        token,
+        duration,
+        interest_rate_monthly
+    ):
+        self.token = token
+        self.duration = duration
+        self.interest_rate_monthly = interest_rate_monthly 
+
+    def sql(self, table_query, aux_table_query): 
+        token_holder_table = table_query
+        safe_owner_table = aux_table_query
+        
+        return  f"""
+            with compound_parameters as (select th._block_number, so.owner, th.safe, th.balance_uint64,
+                lag(th.balance_uint64) over (partition by th.safe order by th._block_number asc NULLS LAST) as old_balance,
+                th.balance_uint64 - lag(th.balance_uint64) over (partition by th.safe order by th._block_number asc NULLS LAST) as change,
```

Comment:
lag(column, 1, 0) removes the need for a coalesce later, the defaults are 1 and NULL but we know we can treat a missing value as 0